File: src/main/java/net/orandja/vw/mods/WhitelistedChestBlock/WhitelistedChestBlock.java
package net.orandja.vw.mods.WhitelistedChestBlock;

import net.minecraft.block.BlockState;
import net.minecraft.block.ChestBlock;
import net.minecraft.block.enums.ChestType;
import net.minecraft.entity.Entity;
import net.minecraft.entity.player.PlayerEntity;
import net.minecraft.inventory.CraftingInventory;
import net.minecraft.item.ItemPlacementContext;
import net.minecraft.item.ItemStack;
import net.minecraft.nbt.NbtCompound;
import net.minecraft.nbt.NbtElement;
import net.minecraft.nbt.NbtList;
import net.minecraft.nbt.NbtString;
import net.minecraft.network.packet.s2c.play.BlockUpdateS2CPacket;
import net.minecraft.recipe.Ingredient;
import net.minecraft.recipe.ShapelessRecipe;
import net.minecraft.server.network.ServerPlayerEntity;
import net.minecraft.text.Text;
import net.minecraft.util.ActionResult;
import net.minecraft.util.Identifier;
import net.minecraft.util.collection.DefaultedList;
import net.minecraft.util.math.BlockPos;
import net.minecraft.world.World;
import net.orandja.mcutils.InventoryUtils;
import net.orandja.mcutils.NBTUtils;
import net.orandja.vw.VW;
import net.orandja.vw.logic.CustomRecipe;
import net.orandja.vw.mods.ProtectBlock.ProtectBlock;
import org.spongepowered.asm.mixin.injection.callback.CallbackInfoReturnable;

import java.util.List;
import java.util.Optional;
import java.util.UUID;
import java.util.stream.Collectors;

public interface WhitelistedChestBlock {

    class WhitelistedChestRecipe extends ShapelessRecipe {

        public WhitelistedChestRecipe(Identifier id, String group, ItemStack output, DefaultedList<Ingredient> ingredients) {
            super(id, group, output, ingredients);
        }

        @Override
        public boolean matches(CraftingInventory craftingInventory, World world) {
            return InventoryUtils.hasListeners(craftingInventory) && InventoryUtils.getPlayer(craftingInventory) != null && super.matches(craftingInventory, world);
        }

        @Override
        public ItemStack craft(CraftingInventory craftingInventory) {
            PlayerEntity player = InventoryUtils.getPlayer(craftingInventory);
            if(player != null) {
                ItemStack chestOutput = getOutput().copy();

                NbtCompound tag = chestOutput.getOrCreateNbt();

                NbtList whitelist = NBTUtils.getOrCreate(tag, "whitelist", (key) -> tag.getList(key, NbtElement.STRING_TYPE), NbtList::new);
                Optional<String> first = whitelist.stream().map(NbtElement::asString).filter(player.getUuidAsString()::equals).findFirst();

                if(first.isPresent()) {
                    String uuid = first.get();
                    whitelist.add(NbtString.of(uuid));


                    NbtCompound display = NBTUtils.getOrCreate(tag, "display", tag::getCompound, NbtCompound::new);
                    NbtList lore = NBTUtils.getOrCreate(display, "Lore", key -> display.getList(key, NbtElement.STRING_TYPE), () -> {
                        NbtList list = new NbtList();
                            list.add(NbtString.of("{\"text\":\"Whitelisted Chest allowed for: \",\"color\":\"green\"}"));
                        return list;
                    });
                    lore.add(NbtString.of("{\"text\":\"— "+ player.getName().getString() +"\",\"color\":\"green\"}"));
                }

                return chestOutput;
            }
            return ItemStack.EMPTY;
        }
    }

    static void addNbt(List<String> list, NbtCompound nbt) {
        if(nbt.contains("whitelist")) {
            nbt.getList("whitelist", NbtElement.STRING_TYPE).stream().map(NbtElement::asString).forEach(list::add);
        }
    }

    static void toNBT(List<String> list, NbtCompound nbt) {
        if(list.size() > 0) {
            NbtList nbtList = new NbtList();
            list.stream().map(NbtString::of).forEach(nbtList::add);
            nbt.put("whitelist", nbtList);
        }
    }

    static String UUIDToUsername(String uuid) {
        //noinspection OptionalGetWithoutIsPresent
        return VW.Companion.getServerReference().get().getUserCache().getByUuid(UUID.fromString(uuid)).get().getName();
    }

    static boolean hasWhitelist(ItemStack stack) {
        //noinspection ConstantConditions
        return stack.hasNbt() && stack.getNbt().contains("whitelist");
    }

    static void sendBlockUpdate(ServerPlayerEntity player, World world, BlockPos... pos) {
        for (BlockPos p : pos) {
            player.networkHandler.sendPacket(new BlockUpdateS2CPacket(world, p));
        }
    }

    static BlockPos getAdjascentPos(BlockState state, BlockPos pos) {
        ChestType chestType = state.get(ChestBlock.CHEST_TYPE);
        if(chestType == ChestType.SINGLE) {
            return null;
        }

        return switch (chestType) {
            case LEFT -> pos.offset(state.get(ChestBlock.FACING).rotateYClockwise());
            case RIGHT -> pos.offset(state.get(ChestBlock.FACING).rotateYCounterclockwise());
            default -> pos.offset(state.get(ChestBlock.FACING));
        };
    }

    static WhitelistedChestBlock getChest(World world, BlockPos pos) {
        try {
            return !world.isClient ? ((WhitelistedChestBlock)world.getBlockEntity(pos)) : null;
        } catch(Exception e) {
            return null;
        }
    }

    static void beforeLaunch() {
        CustomRecipe.Companion.getCustomShapelessRecipes().put(new Identifier("vanillawaifu", "whitelisted_chest"), WhitelistedChestRecipe::new);
        ProtectBlock.DESTROY.add((world, pos, entity) -> {
           WhitelistedChestBlock chest = getChest(world, pos);
           if(chest == null) {
               return true;
           }

           return chest.canDestroy(entity);
        });
        ProtectBlock.EXPLOSION_PROTECTION.add((world, pos) -> {
            WhitelistedChestBlock chest = getChest(world, pos);
            if(chest == null) {
                return false;
            }

            return !chest.hasNoWhitelist();
        });


        ProtectBlock.EXTRACTION_PREVENTION.add((world, pos) -> {
            WhitelistedChestBlock chest = getChest(world, pos);
            if(chest == null) {
                return true;
            }

            return chest.hasWhitelist();
        });
    }

    List<String> getWhitelist();

    default void loadWhitelist(NbtCompound nbt) {
        addNbt(getWhitelist(), nbt);
    }

    default void saveWhitelist(NbtCompound nbt) {
        toNBT(getWhitelist(), nbt);
    }

    default boolean hasWhitelist() {
        return getWhitelist().size() > 0;
    }

    default boolean hasNoWhitelist() {
        return getWhitelist().size() == 0;
    }

    default boolean canDestroy(Entity entity) {
        return hasNoWhitelist() || isWhitelisted(entity);
    }

    default boolean isWhitelisted(Entity entity) {
        if(entity instanceof PlayerEntity player) {
            return isWhitelisted(player);
        } else {
            return false;
        }
    }

    default boolean isWhitelisted(PlayerEntity player) {
        return hasNoWhitelist() || player.isCreative() || getWhitelist().contains(player.getUuidAsString());
    }

    default boolean compareStack(ItemStack stack) {
        //noinspection ConstantConditions
        if(stack.hasNbt() && stack.getNbt().contains("whitelist")) {
            return stack.getNbt().getList("whitelist", NbtElement.STRING_TYPE).stream().map(NbtElement::asString).filter(getWhitelist()::contains).count() == getWhitelist().size();

        }

        return false;
    }

    default void notifyPlayer(PlayerEntity player, CallbackInfoReturnable<ActionResult> info) {
        if(!isWhitelisted(player)) {
            String players = getWhitelist().stream().map(WhitelistedChestBlock::UUIDToUsername).collect(Collectors.joining(","));
            player.sendMessage(Text.of("[{\"text\":\"Whitelisted for: \",\"color\":\"red\"},{\"text\":\""+ players +"\",\"color\":\"green\"}]"), true);
            info.setReturnValue(ActionResult.SUCCESS);
        }
    }

    default void handleStack(ItemStack stack) {
        //noinspection ConstantConditions
        if(stack.hasNbt() && stack.getNbt().contains("whitelist")) {
            stack.getNbt().getList("whitelist", NbtElement.STRING_TYPE).stream().map(NbtElement::asString).forEach(getWhitelist()::add);
        }
    }

    default void onBlockUse(World world, BlockPos pos, PlayerEntity player, CallbackInfoReturnable<ActionResult> info) {
        WhitelistedChestBlock chest = getChest(world, pos);
        if(chest != null) {
            chest.notifyPlayer(player, info);
        }
    }

    default void onBlockPlaced(World world, BlockPos pos, ItemStack stack) {
        WhitelistedChestBlock chest = getChest(world, pos);
        if(chest != null) {
            chest.handleStack(stack);
        }
    }

    default void onPlacementState(ItemPlacementContext ctx, CallbackInfoReturnable<BlockState> info) {
        if(ctx.getWorld().isClient) {
            return;
        }

        BlockPos adjacentPos = getAdjascentPos(info.getReturnValue(), ctx.getBlockPos());
        if(adjacentPos == null) {
            return;
        }
        WhitelistedChestBlock adjacentChest = getChest(ctx.getWorld(), adjacentPos);
        if(adjacentChest == null) {
            return;
        }

        if((hasWhitelist(ctx.getStack()) && adjacentChest.compareStack(ctx.getStack())) || !adjacentChest.hasWhitelist()) {
            return;
        }

        if(ctx.getPlayer() instanceof ServerPlayerEntity player) {
            sendBlockUpdate(player, ctx.getWorld(), ctx.getBlockPos(), adjacentPos);
        }
        info.setReturnValue(info.getReturnValue().with(ChestBlock.CHEST_TYPE, ChestType.SINGLE));
    }
}

File: src/main/kotlin/net/orandja/vw/logic/WhitelistedChestBlock.kt
package net.orandja.vw.logic

import net.minecraft.block.BlockState
import net.minecraft.block.ChestBlock
import net.minecraft.block.enums.ChestType
import net.minecraft.entity.Entity
import net.minecraft.entity.player.PlayerEntity
import net.minecraft.inventory.CraftingInventory
import net.minecraft.item.ItemPlacementContext
import net.minecraft.item.ItemStack
import net.minecraft.nbt.NbtCompound
import net.minecraft.nbt.NbtElement
import net.minecraft.nbt.NbtList
import net.minecraft.nbt.NbtString
import net.minecraft.network.packet.s2c.play.BlockUpdateS2CPacket
import net.minecraft.recipe.Ingredient
import net.minecraft.recipe.ShapelessRecipe
import net.minecraft.server.network.ServerPlayerEntity
import net.minecraft.text.Text
import net.minecraft.util.ActionResult
import net.minecraft.util.Identifier
import net.minecraft.util.collection.DefaultedList
import net.minecraft.util.math.BlockPos
import net.minecraft.world.World
import net.orandja.mcutils._addAll
import net.orandja.mcutils.getOrCreate
import net.orandja.mcutils.getPlayer
import net.orandja.mcutils.hasListeners
import net.orandja.vw.VW
import org.spongepowered.asm.mixin.injection.callback.CallbackInfo
import org.spongepowered.asm.mixin.injection.callback.CallbackInfoReturnable
import java.util.*

private fun ArrayList<String>.addNbt(nbt: NbtCompound) {
    if (nbt.contains("whitelist"))
        nbt.getList("whitelist", 8).map(NbtElement::asString).forEach(::add)
}

private fun ArrayList<String>.toNBT(nbt: NbtCompound) {
    if (size > 0)
        nbt.put("whitelist", NbtList()._addAll(map(NbtString::of)))
}

private fun String.asUserName(): String {
    return VW.serverReference.get().userCache.getByUuid(UUID.fromString(this)).get().name
}

private fun ItemStack.hasWhiteList(): Boolean {
    return nbt?.contains("whitelist") ?: false
}

private fun ServerPlayerEntity.sendBlockUpdate(world: World, vararg pos: BlockPos) {
    pos.forEach { networkHandler.sendPacket(BlockUpdateS2CPacket(world, it)) }
}

private fun BlockState.getAdjacentPos(pos: BlockPos): BlockPos? {
    val chestType = get(ChestBlock.CHEST_TYPE)
    if (chestType == ChestType.SINGLE) {
        return null
    }

    return pos.offset(get(ChestBlock.FACING).let {
        when (chestType) {
            ChestType.RIGHT -> it.rotateYCounterclockwise()
            ChestType.LEFT -> it.rotateYClockwise()
            else -> it
        }
    })
}

interface WhitelistedChestBlock {

    companion object {
        fun getChest(world: World, pos: BlockPos): WhitelistedChestBlock? {
            return if (!world.isClient) world.getBlockEntity(pos) as? WhitelistedChestBlock else null
        }

        fun beforeLaunch() {
            CustomRecipe.customShapelessRecipes[Identifier("vanillawaifu", "whitelisted_chest")] =
                ::WhitelistedChestRecipe
            ProtectBlock.DESTROY.add { world, pos, entity -> getChest(world, pos)?.canDestroy(entity) ?: true }
            ProtectBlock.EXPLOSION_PROTECTION.add { world, pos -> getChest(world, pos)?.hasNoWhitelist() == false }
            ProtectBlock.EXTRACTION_PREVENTION.add { world, pos -> getChest(world, pos)?.hasWhitelist() == true }
        }
    }

    var whitelist: ArrayList<String>

    fun loadWhitelist(nbt: NbtCompound) {
        whitelist.addNbt(nbt)
    }
    fun saveWhitelist(nbt: NbtCompound) {
        whitelist.toNBT(nbt)
    }
    fun hasWhitelist(): Boolean {
        return whitelist.size > 0
    }
    fun hasNoWhitelist(): Boolean {
        return whitelist.size == 0
    }

    fun canDestroy(player: Entity): Boolean {
        return !hasWhitelist() || isWhitelisted(player as? PlayerEntity ?: return false)
    }

    private fun isWhitelisted(player: PlayerEntity): Boolean =
        !hasWhitelist() || player.isCreative || whitelist.contains(player.uuidAsString)

    private fun compareStack(stack: ItemStack): Boolean {
        return (stack.nbt?.getList("whitelist", 8)?.map(NbtElement::asString)
            ?: return false).count { whitelist.contains(it) } == whitelist.size
    }

    private fun handlePlayer(player: PlayerEntity, info: CallbackInfoReturnable<ActionResult>) {
        if (!isWhitelisted(player)) {
            player.sendMessage(
                Text.Serializer.fromJson(
                    """[{"text":"Whitelisted for: ","color":"red"},{"text":"${
                        whitelist.joinToString(
                            ",",
                            transform = String::asUserName
                        )
                    }","color":"green"}]"""
                ), true
            )
            info.returnValue = ActionResult.SUCCESS
        }
    }

    private fun handleStack(stack: ItemStack) {
        stack.nbt?.getList("whitelist", 8)?.map(NbtElement::asString)?.forEach(whitelist::add)
    }

    fun onBlockUse(world: World, pos: BlockPos, player: PlayerEntity, info: CallbackInfoReturnable<ActionResult>) {
        getChest(world, pos)?.handlePlayer(player, info)
    }

    fun onBlockPlaced(world: World, pos: BlockPos, stack: ItemStack, info: CallbackInfo) {
        getChest(world, pos)?.handleStack(stack)
    }

    fun onPlacementState(ctx: ItemPlacementContext, info: CallbackInfoReturnable<BlockState>) {
        if (ctx.world.isClient)
            return

        val adjacentPos = info.returnValue.getAdjacentPos(ctx.blockPos) ?: return
        val adjacentChest = getChest(ctx.world, adjacentPos)

        if ((ctx.stack.hasWhiteList() && adjacentChest?.compareStack(ctx.stack) == true) || adjacentChest?.hasWhitelist() == false) {
            return
        }

        (ctx.player as? ServerPlayerEntity)?.sendBlockUpdate(ctx.world, ctx.blockPos, adjacentPos)
        info.returnValue = info.returnValue.with(ChestBlock.CHEST_TYPE, ChestType.SINGLE)
    }
}

class WhitelistedChestRecipe(id: Identifier, group: String, output: ItemStack, ingredients: DefaultedList<Ingredient>) :
    ShapelessRecipe(id, group, output, ingredients) {

    override fun matches(inventory: CraftingInventory, world: World): Boolean {
        return inventory.hasListeners() && inventory.getPlayer() != null && super.matches(inventory, world)
    }

    override fun craft(inventory: CraftingInventory): ItemStack {
        val player = inventory.getPlayer() ?: return ItemStack.EMPTY
        if (inventory.hasListeners()) {
            val chestOutput = output.copy()

            chestOutput.orCreateNbt.apply {
                val whitelist =
                    this.getOrCreate("whitelist", { this.getList(it, 8) }, ::NbtList) ?: return ItemStack.EMPTY
                if (whitelist.map(NbtElement::asString).find(player.uuidAsString::equals) == null) {
                    whitelist.add(NbtString.of(player.uuidAsString))

                    val display = this.getOrCreate("display", this::get, ::NbtCompound) as NbtCompound
                    val list = display.getOrCreate("Lore", display::get) {
                        val lore = NbtList()
                        lore.add(NbtString.of("""{"text":"Whitelisted Chest allowed for: ","color":"green"}"""))
                        lore
                    } as NbtList

                    list.add(NbtString.of("""{"text":"— ${player.name.string}","color":"green"}"""))
                }
            }

            return chestOutput
        }
        return ItemStack.EMPTY
    }
}